src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1274,83 +1456,44 @@ private void WriteHeaders(HttpRequestMessage request, ref ArrayBuffer headerBuff
             }
         }
 
-        [DoesNotReturn]
-        private void ThrowShutdownException()
-        {
-            Debug.Assert(Monitor.IsEntered(SyncObject));
-
-            if (_abortException != null)
-            {
-                // We had an IO failure on the connection. Don't retry in this case.
-                throw new HttpRequestException(SR.net_http_client_execution_error, _abortException);
-            }
-
-            // Connection is being gracefully shutdown. Allow the request to be retried.
-            Exception innerException;
-            if (_lastStreamId != -1)
-            {
-                // We must have received a GOAWAY.
-                innerException = new IOException(SR.net_http_server_shutdown);
-            }
-            else
-            {
-                // We must either be disposed or out of stream IDs.
-                Debug.Assert(_disposed || _nextStream == MaxStreamId);
-
-                innerException = new ObjectDisposedException(nameof(Http2Connection));
-            }
-
-            ThrowRetry(SR.net_http_client_execution_error, innerException);
-        }
-
-        private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request, CancellationToken cancellationToken, bool mustFlush)
+        private void AddStream(Http2Stream http2Stream)
         {
-            // Enforce MAX_CONCURRENT_STREAMS setting value.  We do this before anything else, e.g. renting buffers to serialize headers,
-            // in order to avoid consuming resources in potentially many requests waiting for access.
-            try
+            lock (SyncObject)
             {
-                if (!_concurrentStreams.TryRequestCreditNoWait(1))
+                if (_nextStream == MaxStreamId)
```

Comment:
I see that this part roughly corresponds to the original `ThrowShutdownException`, so my question is: should we check `_abortException` here?

Reply:
If `_abortException` is set, then `_shutdown` is true. We check `_shutdown` just below, and if it's true, we throw a retryable exception.